bson/value_reader.go

```diff
@@ -46,69 +72,81 @@ var vrPool = sync.Pool{
 
 // valueReader is for reading BSON values.
 type valueReader struct {
-	r      *bufio.Reader
+	src    valueReaderByteSrc
 	offset int64
 
 	stack []vrState
 	frame int64
 }
 
-func getDocumentReader(r io.Reader) *valueReader {
-	vr := vrPool.Get().(*valueReader)
-
-	vr.offset = 0
-	vr.frame = 0
-
-	vr.stack = vr.stack[:1]
-	vr.stack[0] = vrState{mode: mTopLevel}
-
-	br := bufioReaderPool.Get().(*bufio.Reader)
-	br.Reset(r)
-	vr.r = br
-
-	return vr
+func getBufferedDocumentReader(b []byte) *valueReader {
+	return newBufferedDocumentReader(b)
 }
 
-func putDocumentReader(vr *valueReader) {
+func putBufferedDocumentReader(vr *valueReader) {
 	if vr == nil {
 		return
 	}
 
-	bufioReaderPool.Put(vr.r)
-	vr.r = nil
+	vr.src.reset()
+
+	// Reset src and stack to avoid holding onto memory.
+	vr.src = nil
+	vr.frame = 0
+	vr.stack = vr.stack[:0]
 
 	vrPool.Put(vr)
 }
 
 // NewDocumentReader returns a ValueReader using b for the underlying BSON
 // representation.
 func NewDocumentReader(r io.Reader) ValueReader {
-	return newDocumentReader(r)
-}
-
-// newValueReader returns a ValueReader that starts in the Value mode instead of in top
-// level document mode. This enables the creation of a ValueReader for a single BSON value.
-func newValueReader(t Type, r io.Reader) ValueReader {
-	vr := newDocumentReader(r)
-	if len(vr.stack) == 0 {
-		vr.stack = make([]vrState, 1, 5)
-	}
-	vr.stack[0].mode = mValue
-	vr.stack[0].vType = t
-	return vr
-}
-
-func newDocumentReader(r io.Reader) *valueReader {
 	stack := make([]vrState, 1, 5)
 	stack[0] = vrState{
 		mode: mTopLevel,
 	}
+
 	return &valueReader{
-		r:     bufio.NewReader(r),
+		src:   &streamingValueReader{br: bufio.NewReader(r), offset: 0},
 		stack: stack,
 	}
 }
 
+// newBufferedValueReader returns a ValueReader that starts in the Value mode
+// instead of in top level document mode. This enables the creation of a
+// ValueReader for a single BSON value.
+func newBufferedValueReader(t Type, b []byte) ValueReader {
+	bVR := newBufferedDocumentReader(b)
+
+	bVR.stack[0].vType = t
+	bVR.stack[0].mode = mValue
+
+	return bVR
+}
+
+func newBufferedDocumentReader(b []byte) *valueReader {
+	vr := vrPool.Get().(*valueReader)
+
+	vr.src = &bufferedValueReader{}
+	vr.src.(*bufferedValueReader).buf = b
+	vr.src.(*bufferedValueReader).offset = 0
```

Comment:
Optional: Suggested simplification.

```suggestion
	vr.src = &bufferedValueReader{
		buf:    b,
		offset: 0,
	}
```